### PR TITLE
[libc++] Check correctly ref-qualified __is_callable in algorithms

### LINKED ISSUE #69554: <algorithm>: `__is_callable` checks whether the callable can be called with rvalue
https://github.com/llvm/llvm-project/blob/a574242f196dd0afaa2c591d6cc5716fd67bc738/libcxx/include/__type_traits/is_callable.h#L22-L29

Some std algorithms check whether callable is an ordinary function through `__is_callable`. However, according to its above definition, this seems to only check whether callable can be called with an rvalue, which seems to be an oversight.

```cpp
#include <algorithm>

struct Cmp {
  bool operator()(int, int) const &;
  bool operator()(int, int) && = delete;
};

int main() {
  return std::minmax({42, 0, -42}, Cmp{}).first;
}
```
https://godbolt.org/z/8a45rPdqo

### PR DESCRIPTION
**Pull request by changkhothuychung**
We were only checking that the comparator was rvalue callable,
when in reality the algorithms always call comparators as lvalues.
This patch also refactors the tests for callable requirements and
expands it to a few missing algorithms.

Fixes #69554

### CODE REVIEW

**Comment by changkhothuychung** [comment]
@ldionne ldi
can you give me a test example so I can add for this PR? should it be a verify or pass test?

**Review by ldionne — CHANGES_REQUESTED** [review]

**Review by ldionne — CHANGES_REQUESTED** [review]
Also, please add a description of the patch in the PR description!

**Review by EricWF — CHANGES_REQUESTED** [review]
I don't understand the motivation behind static asserts like this.  They have a number of downsides.

1.  They cost compile time and memory. The instantiations they perform live in the compilers AST until the end of the TU.
2. They don't catch broken code. The broken code is already broken, this just diagnoses it differently.
3.  They break well formed code.

As evidenced here, when we get the needless checks wrong, they break valid code. Which is a lot worse than providing a slightly better diagnostic to ill-formed code.

I say we get rid of them entirely.

**Comment by changkhothuychung** [comment]
> I don't understand the motivation behind static asserts like this. They have a number of downsides.
> 
> 1. They cost compile time and memory. The instantiations they perform live in the compilers AST until the end of the TU.
> 2. They don't catch broken code. The broken code is already broken, this just diagnoses it differently.
> 3. They break well formed code.
> 
> As evidenced here, when we get the needless checks wrong, they break valid code. Which is a lot worse than providing a slightly better diagnostic to ill-formed code.
> 
> I say we get rid of them entirely.

@ldionne l
do you have any opinions on this? I will wait for your response before I continue.

File: libcxx/test/libcxx/algorithms/const_callable.veriry.cpp
Comment (ldionne) [inline libcxx/test/libcxx/algorithms/const_callable.veriry.cpp:1]: Missing license.

File: libcxx/test/libcxx/algorithms/const_callable.veriry.cpp
Comment (ldionne) [inline libcxx/test/libcxx/algorithms/const_callable.veriry.cpp:16]: Missing newline at the end of the file.

File: libcxx/test/libcxx/algorithms/const_callable.veriry.cpp
Comment (ldionne) [inline libcxx/test/libcxx/algorithms/const_callable.veriry.cpp:1]: Also, there is a typo in the file name (`.veriry.cpp`).

**Comment by ldionne** [comment]
> do you have any opinions on this? I will wait for your response before I continue.

@philnik777 Do you remember why you added the `__is_callable` checks in the first place? The Standard isn't all that clear about what the requirements are for classic algorithms, so I think removing these checks would also be a viable path forward. I seem to remember there was a good reason for adding those, so I'm reluctant to do it without remembering the context.

**Comment by philnik777** [comment]
> > do you have any opinions on this? I will wait for your response before I continue.
> 
> @philnik777 Do you remember why you added the `__is_callable` checks in the first place? The Standard isn't all that clear about what the requirements are for classic algorithms, so I think removing these checks would also be a viable path forward. I seem to remember there was a good reason for adding those, so I'm reluctant to do it without remembering the context.

We added them to disallow the use of functions to member pointers etc. to be passed to the classic algorithms.

**Comment by ldionne** [comment]
> > @philnik777 Do you remember why you added the `__is_callable` checks in the first place? The Standard isn't all that clear about what the requirements are for classic algorithms, so I think removing these checks would also be a viable path forward. I seem to remember there was a good reason for adding those, so I'm reluctant to do it without remembering the context.
> 
> We added them to disallow the use of functions to member pointers etc. to be passed to the classic algorithms.

Oh yeah, that's it. Wouldn't that fail to compile already though?

File: libcxx/test/libcxx/random/random.uniform.real/test.cpp
Comment (changkhothuychung) [reply libcxx/test/libcxx/random/random.uniform.real/test.cpp]: @ldionne l
did you want to make the second operator `const &` ? I am getting errors, overloading either requires both functions to have ref-qualifiers or both lack.

Comment (changkhothuychung) [reply libcxx/test/libcxx/random/random.uniform.real/test.cpp]: I added a `.pass` file, can you take a look and let me know if they are good? if yes, I will add similar tests for the remaining algorithms.

**Comment by changkhothuychung** [comment]
@ldionne it looks like there is a bug with the use of `std::lower_bound(begin(collatenames), end(collatenames), s, use_strcmp());` in the file regex.cpp, because `use_strcmp` doesn't have a const reference operator overloading. Should we add one to that code?

**Comment by ldionne** [comment]
> @ldionne it looks like there is a bug with the use of `std::lower_bound(begin(collatenames), end(collatenames), s, use_strcmp());` in the file regex.cpp, because `use_strcmp` doesn't have a const reference operator overloading. Should we add one to that code?

Yes, we should add a const-qualification to `use_strcmp` in our own code.

@EricWF Now that I think about it some more, I think the issue these static asserts are addressing is that without them, we'll end up using `std::__invoke` to call the functions, which means that passing a PMF as a function would appear to work with classic algorithms, when in reality it's non-standard. So these static assertions are guarding us against Hyrum's law since our underlying algorithm implementation would allow for PMFs to be passed.

**Review by ldionne — APPROVED** [review]
Will merge once CI is green.

**Comment by vitalybuka** [comment]
This one breaks multiple build bot 

https://lab.llvm.org/buildbot/#/builders/66/builds/2419

```
/b/sanitizer-x86_64-linux/build/build_default/runtimes/runtimes-bins/compiler-rt/lib/sanitizer_common/symbolizer/RTSanitizerCommonSymbolizerInternal.i386/symbolizer/libcxx/include/c++/v1/__algorithm/upper_bound.h:54:17: error: static assertion failed due to requirement '__is_callable<(lambda at /b/sanitizer-x86_64-linux/build/llvm-project/llvm/lib/DebugInfo/CodeView/LazyRandomTypeCollection.cpp:174:33) &, const llvm::codeview::TypeIndexOffset &, const llvm::codeview::TypeIndex &>::value': The comparator has to be callable
   54 |   static_assert(__is_callable<_Compare&, decltype(*__first), const _Tp&>::value, "The comparator has to be callable");
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/b/sanitizer-x86_64-linux/build/llvm-project/llvm/include/llvm/ADT/STLExtras.h:1974:15: note: in instantiation of function template specialization 'std::upper_bound<llvm::FixedStreamArrayIterator<llvm::codeview::TypeIndexOffset>, llvm::codeview::TypeIndex, (lambda at /b/sanitizer-x86_64-linux/build/llvm-project/llvm/lib/DebugInfo/CodeView/LazyRandomTypeCollection.cpp:174:33)>' requested here
 1974 |   return std::upper_bound(adl_begin(Range), adl_end(Range),
      |               ^
1 error generated.
[191/638] Building CXX object lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/TypeTableCollection.cpp.o
```

**Comment by ldionne** [comment]
I'll revert. I think that's a bug in the order of arguments I passed to `__is_callable` for `upper_bound`. That wasn't intended.

**Pull request merged**
